pkg/apis/certificates/validation/validation.go

```diff
@@ -140,8 +196,128 @@ func ValidateCertificateSigningRequestSignerName(fldPath *field.Path, signerName
+}
+
+func validateCertificateSigningRequestUpdate(newCSR, oldCSR *certificates.CertificateSigningRequest, opts certificateValidationOptions) field.ErrorList {
+	validationErrorList := validateCertificateSigningRequest(newCSR, opts)
 	metaUpdateErrorList := apivalidation.ValidateObjectMetaUpdate(&newCSR.ObjectMeta, &oldCSR.ObjectMeta, field.NewPath("metadata"))
+
+	// TODO: prevent removal of existing Approved, Denied, or Failed conditions
+
+	if !opts.allowSettingApproval {
+		// TODO: prevent addition of Approval or Denied conditions
+	}
+
+	if !bytes.Equal(newCSR.Status.Certificate, oldCSR.Status.Certificate) {
+		if !opts.allowSettingCertificate {
+			// TODO: prevent setting certificate
+		} else if !opts.allowResettingCertificate && len(oldCSR.Status.Certificate) > 0 {
+			// TODO: prevent resetting certificate
+		}
+	}
+
 	return append(validationErrorList, metaUpdateErrorList...)
 }
+
+// getValidationOptions returns the validation options to be
+// compatible with the specified version and existing CSR.
+// oldCSR may be nil if this is a create request.
+// validation options related to subresource-specific capabilities are set to false.
+func getValidationOptions(version schema.GroupVersion, oldCSR *certificates.CertificateSigningRequest) certificateValidationOptions {
+	return certificateValidationOptions{
+		version:                      version,
+		allowResettingCertificate:    allowResettingCertificate(version),
+		allowBothApprovedAndDenied:   allowBothApprovedAndDenied(oldCSR),
+		allowLegacySignerName:        allowLegacySignerName(version, oldCSR),
+		allowDuplicateConditionTypes: allowDuplicateConditionTypes(version, oldCSR),
+		allowEmptyConditionType:      allowEmptyConditionType(version, oldCSR),
+	}
+}
+
+func allowResettingCertificate(version schema.GroupVersion) bool {
+	// compatibility with v1beta1
+	return version == certificatesv1beta1.SchemeGroupVersion
+}
+
+func allowBothApprovedAndDenied(oldCSR *certificates.CertificateSigningRequest) bool {
+	if oldCSR == nil {
+		return false
+	}
+	approved := false
+	denied := false
+	for _, c := range oldCSR.Status.Conditions {
+		if c.Type == certificates.CertificateApproved {
+			approved = true
+		} else if c.Type == certificates.CertificateDenied {
+			denied = true
+		}
+	}
+	// compatibility with existing data
+	return approved && denied
+}
+
+func allowLegacySignerName(version schema.GroupVersion, oldCSR *certificates.CertificateSigningRequest) bool {
+	switch {
+	case version == certificatesv1beta1.SchemeGroupVersion:
+		return true // compatibility with v1beta1
+	case oldCSR != nil && oldCSR.Spec.SignerName == certificates.LegacyUnknownSignerName:
+		return true // compatibility with existing data
+	default:
+		return false
+	}
+}
+
+func allowDuplicateConditionTypes(version schema.GroupVersion, oldCSR *certificates.CertificateSigningRequest) bool {
+	switch {
+	case version == certificatesv1beta1.SchemeGroupVersion:
+		return true // compatibility with v1beta1
```

Comment:
github doesn't have a wide enough pallet of emojis for me to properly express my reaction here